File: distroquiz.pl
#the Perl 6 version is way prettier :)

{
    package Question;
    use Moo;
    has question => ( is => 'ro' );
    has yes      => ( is => 'ro' );
    has no       => ( is => 'ro' );
}

use 5.012;
use JSON::XS;
use File::Slurp 'slurp';
use encoding 'utf8';

my $data = decode_json slurp 'db.json';
my %answers;
my %questions;

while (my ($k, $v) = each %{$data->{'answers'}}) {
    $answers{$k} = $v;
}

while (my ($k, $v) = each %{$data->{'questions'}}) {
    $questions{$k} = Question->new(
        question => $v->{'question'},
        yes      => $v->{'yes'}[0] eq 'answer'
                     ? sub { $answers{$v->{'yes'}[1]} }
                     : sub { $questions{$v->{'yes'}[1]} },
        no       => $v->{'no'}[0] eq 'answer'
                     ? sub { $answers{$v->{'no'}[1]} }
                     : sub { $questions{$v->{'no'}[1]} },
    );
}

my $current = $questions{$data->{'root'}};

while (ref $current ne 'ARRAY') {
    print $current->question, " ";
    if (<> ~~ /tak/) {
        $current = $current->yes->();
    } else {
        $current = $current->no->();
    }
}

say "Dystrybucje dla Ciebie to mniej-więcej: ", join(', ', @$current);

File: distroquiz.rb
require 'json'

dbfile = 'db.json'

class Question
    attr_accessor :question
    attr_accessor :yes
    attr_accessor :no

    def initialize(question, yes, no)
        @question = question
        @yes      = yes
        @no       = no
    end
end

answers   = {}
questions = {}

data = JSON.parse(IO.read(dbfile))

data['answers'].each do |k, v|
    answers[k] = v
end

data['questions'].each do |k, v|
    yes = v['yes'][0] == "answer" ? lambda { answers[v['yes'][1]]   } \
                                  : lambda { questions[v['yes'][1]] }
    no  = v['no'][0]  == "answer" ? lambda { answers[v['no'][1]]    } \
                                  : lambda { questions[v['no'][1]]  }
    questions[k] = Question.new(v['question'], yes, no)
end

current = questions[data['root']]

while current.is_a?(Question)
    print current.question, ' '
    ans = gets
    if ans.match(/tak/)
        current = current.yes.call
    else
        current = current.no.call
    end
end

puts "Dystrubucje dla Ciebie to mniej-wiecej: ", current.join(', ')
